Fetch device info by vendor id and vendor profile id

vlasebian opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary

<!--
A short summary, referencing related issues:
References #0000, etc.
Pull requests may close issues (using Closes #0000) only for minor changes not needing tests on a staging environment.
Typically, issues should be closed manually after validation.
-->
References #7222 

#### Changes

- ~~Introduce a new EndDeviceProvisioningInfo message~~
- ~~Introduce a new endpoint, `/dr/vendors/{vendor_id}/profiles/{vendor_profile_id}/provisioning-info`~~
- ~~Implement bleve index for EndDeviceProvisioningInfo~~
- ~~Implement a new method in the device repository to list end device profiles by brand id~~
- ~~Added tests for the remote store and bleve store~~

After the discussion in this PR, I dropped all the changes related to the GetProvisioningInfo endpoint and started fresh. These are the changes that I did:

- Dropped the profiles index from the bleve store (it was used only to fetch templates)
- Added a new index for templates in the bleve store, so templates are returned directly instead of fetching a profile and then building the template
- Updated the reading of profiles from the device repository to account for the profile identifiers in the vendor index too. This is how reading end device profiles works now:
- All the profiles are read as previously (all the device models are checked for profile identifiers, then the profiles are read from their own yaml files)
- The vendor index is checked for any profile identifiers. For each profile identifier, the profile corresponding to that identifier is enriched with the vendor profile id. This is needed for the bleve index to be able to search using vendor id + vendor profile id.
- Dropped the profile parameter from the GetTemplate method in the store interface as it's not needed anymore
- Updated tests, fixed linter warnings

#### Testing

For testing, I did curl requests to check all the endpoint bindings for GetTemplate.

##### Steps

Prerequisites:
- Have tts built and running locally
- Have an application created
- Get a bearer token to authenticate the request (can be done by opening the console, logging in with the default credentials and checking a request header in the network tab of the developer console)

##### Results

1. `curl -s -H "Authorization: $TTN_AUTH_TOKEN" localhost:1885/api/v3/dr/vendors/474/profiles/1/template | jq`

```json
{
  "end_device": {
    "version_ids": {
      "brand_id": "minol-zenner",
      "model_id": "edc-communication-module",
      "hardware_version": "1.0",
      "firmware_version": "2.02.0",
      "band_id": "EU_863_870"
    },
    "lorawan_version": "MAC_V1_0_3",
    "lorawan_phy_version": "PHY_V1_0_3_REV_A",
    "supports_join": true,
    "mac_settings": {
      "supports_32_bit_f_cnt": true
    },
    "formatters": {
      "up_formatter": "FORMATTER_REPOSITORY",
      "down_formatter": "FORMATTER_REPOSITORY"
    }
  },
  "field_mask": {
    "paths": [
      "version_ids",
      "supports_join",
      "supports_class_b",
      "supports_class_c",
      "lorawan_version",
      "lorawan_phy_version",
      "formatters",
      "mac_settings.supports_32_bit_f_cnt"
    ]
  }
}
```

2. `curl -s -H "Authorization: $TTN_AUTH_TOKEN" localhost:1885/api/v3/dr/applications/app1/vendors/474/profiles/1/template | jq`

```json
{
  "end_device": {
    "version_ids": {
      "brand_id": "minol-zenner",
      "model_id": "edc-communication-module",
      "hardware_version": "1.0",
      "firmware_version": "2.02.0",
      "band_id": "EU_863_870"
    },
    "lorawan_version": "MAC_V1_0_3",
    "lorawan_phy_version": "PHY_V1_0_3_REV_A",
    "supports_join": true,
    "mac_settings": {
      "supports_32_bit_f_cnt": true
    },
    "formatters": {
      "up_formatter": "FORMATTER_REPOSITORY",
      "down_formatter": "FORMATTER_REPOSITORY"
    }
  },
  "field_mask": {
    "paths": [
      "version_ids",
      "supports_join",
      "supports_class_b",
      "supports_class_c",
      "lorawan_version",
      "lorawan_phy_version",
      "formatters",
      "mac_settings.supports_32_bit_f_cnt"
    ]
  }
}
```

3. `curl -s -H "Authorization: $TTN_AUTH_TOKEN" localhost:1885/api/v3/dr/brands/the-things-industries/models/generic-node-sensor-edition/1.0/US_902_928/template | jq`

```json
{
  "end_device": {
    "version_ids": {
      "brand_id": "the-things-industries",
      "model_id": "generic-node-sensor-edition",
      "firmware_version": "1.0",
      "band_id": "US_902_928"
    },
    "lorawan_version": "MAC_V1_0_3",
    "lorawan_phy_version": "PHY_V1_0_3_REV_A",
    "supports_join": true,
    "mac_settings": {
      "supports_32_bit_f_cnt": true
    },
    "formatters": {
      "up_formatter": "FORMATTER_REPOSITORY",
      "down_formatter": "FORMATTER_REPOSITORY"
    }
  },
  "field_mask": {
    "paths": [
      "version_ids",
      "supports_join",
      "supports_class_b",
      "supports_class_c",
      "lorawan_version",
      "lorawan_phy_version",
      "formatters",
      "mac_settings.supports_32_bit_f_cnt"
    ]
  }
}
```

4. `curl -s -H "Authorization: $TTN_AUTH_TOKEN" localhost:1885/api/v3/dr/applications/app1/brands/the-things-industries/models/generic-node-sensor-edition/1.0/US_902_928/template | jq`

```json
{
  "end_device": {
    "version_ids": {
      "brand_id": "the-things-industries",
      "model_id": "generic-node-sensor-edition",
      "firmware_version": "1.0",
      "band_id": "US_902_928"
    },
    "lorawan_version": "MAC_V1_0_3",
    "lorawan_phy_version": "PHY_V1_0_3_REV_A",
    "supports_join": true,
    "mac_settings": {
      "supports_32_bit_f_cnt": true
    },
    "formatters": {
      "up_formatter": "FORMATTER_REPOSITORY",
      "down_formatter": "FORMATTER_REPOSITORY"
    }
  },
  "field_mask": {
    "paths": [
      "version_ids",
      "supports_join",
      "supports_class_b",
      "supports_class_c",
      "lorawan_version",
      "lorawan_phy_version",
      "formatters",
      "mac_settings.supports_32_bit_f_cnt"
    ]
  }
}
```

##### Regressions

<!--
Please indicate features that this change could affect and how that was tested.
Also describe the steps required for others to test these regressions.
-->

...

#### Notes for Reviewers

<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

...

#### Checklist

<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Testing: The steps/process to test this feature are clearly explained including testing for regressions.
- [x] Infrastructure: If infrastructural changes (e.g., new RPC, configuration) are needed, a separate issue is created in the infrastructural repositories.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

KrishnaIyer commented on pkg/devicerepository/store/bleve/init.go:
```diff
@@ -203,11 +213,67 @@ func (c Config) Initialize(ctx context.Context, lorawanDevicesPath string, overw
+
+			for _, profile := range profiles.Profiles {
+				vendorID := strconv.Itoa(int(brand.LoraAllianceVendorId))
+				vendorProfileID := strconv.Itoa(int(profile.VendorProfileID))
+
+				template, err := s.GetTemplate(&ttnpb.GetTemplateRequest{
+					EndDeviceProfileIds: &ttnpb.GetTemplateRequest_EndDeviceProfileIdentifiers{
+						VendorId:        brand.LoraAllianceVendorId,
+						VendorProfileId: profile.VendorProfileID,
```
The `VendorProfileID` field is removed from the Profile. This is now instead in the `index.yaml` of each vendor. I think we need to account for this also in the other parts of the code where we assume that `vendorProfileID` is in the profile.

vlasebian replied:
Changed the code to get the profile from the vendor index 👍

vlasebian commented:
Hey, @johanstokking! @KrishnaIyer and I were discussing the final format for the response of this endpoint. As it is implemented now, the proto message is the following:
```proto
message EndDeviceProvisioningInfo {
  EndDeviceModel end_device_model = 1;
  EndDeviceTemplate end_device_template = 2;
}
```
The full response looks like this:
```json
{
  "end_device_model": {
    "name": "Generic Node (Sensor Edition)",
    "description": "The Things Industries Generic Node Sensor Edition is a LoRaWAN® development board that consists of a temperature and humidity sensor and an accelerometer. These onboard sensors can be used to measure temperature, humidity, motion, free fall, and orientation. In addition to that, it provides expansion slots for connecting various sensors. It is suitable for a wide range of use cases that cover industrial, farming, facility management, and leisure.",
    "key_provisioning": [
      "custom"
    ],
    "photos": {
      "main": "gnse.png"
    },
    "product_url": "https://www.genericnode.com/",
    "datasheet_url": "https://github.com/TheThingsIndustries/generic-node-se"
  },
  "end_device_template": {
    "end_device": {
      "version_ids": {
        "brand_id": "the-things-industries",
        "model_id": "generic-node-sensor-edition",
        "hardware_version": "1.1",
        "firmware_version": "1.0",
        "band_id": "EU_863_870"
      },
      "lorawan_version": "MAC_V1_0_3",
      "lorawan_phy_version": "PHY_V1_0_3_REV_A",
      "supports_join": true,
      "mac_settings": {
        "supports_32_bit_f_cnt": true
      },
      "formatters": {
        "up_formatter": "FORMATTER_REPOSITORY",
        "down_formatter": "FORMATTER_REPOSITORY"
      }
    },
    "field_mask": {
      "paths": [
        "version_ids",
        "supports_join",
        "supports_class_b",
        "supports_class_c",
        "lorawan_version",
        "lorawan_phy_version",
        "formatters",
        "mac_settings.supports_32_bit_f_cnt"
      ]
    }
  }
}
```

Our question is, is it relevant to have the full device template response? Wouldn't the device be enough for this case? What I mean is have the proto message as:
```proto
message EndDeviceProvisioningInfo {
  EndDeviceModel end_device_model = 1;
  EndDevice end_device = 2;
}
```
And the full response like this:
```json
{
  "end_device_model": {
    "name": "Generic Node (Sensor Edition)",
    "description": "The Things Industries Generic Node Sensor Edition is a LoRaWAN® development board that consists of a temperature and humidity sensor and an accelerometer. These onboard sensors can be used to measure temperature, humidity, motion, free fall, and orientation. In addition to that, it provides expansion slots for connecting various sensors. It is suitable for a wide range of use cases that cover industrial, farming, facility management, and leisure.",
    "key_provisioning": [
      "custom"
    ],
    "photos": {
      "main": "gnse.png"
    },
    "product_url": "https://www.genericnode.com/",
    "datasheet_url": "https://github.com/TheThingsIndustries/generic-node-se"
  },
  "end_device": {
    "version_ids": {
      "brand_id": "the-things-industries",
      "model_id": "generic-node-sensor-edition",
      "hardware_version": "1.1",
      "firmware_version": "1.0",
      "band_id": "EU_863_870"
    },
    "lorawan_version": "MAC_V1_0_3",
    "lorawan_phy_version": "PHY_V1_0_3_REV_A",
    "supports_join": true,
    "mac_settings": {
      "supports_32_bit_f_cnt": true
    },
    "formatters": {
      "up_formatter": "FORMATTER_REPOSITORY",
      "down_formatter": "FORMATTER_REPOSITORY"
    }
  }
}
```

johanstokking commented:
@vlasebian @KrishnaIyer can you please triage this PR (assignee, milestone) so I know priorities for review.

KrishnaIyer commented:
Yup. @vlasebian: Always make sure to add the milestone and assign yourself to your PRs.

KrishnaIyer commented on pkg/devicerepository/store/remote/schema.go:
```diff
@@ -60,7 +60,20 @@ type VendorsIndex struct {
+	ProfileIDs map[string]VendorProfile `yaml:"profileIDs,omitempty"`
+}
+
+type VendorProfile struct {
```
This is an End Device Profile.

vlasebian replied:
Renamed 👍

KrishnaIyer commented on api/ttn/lorawan/v3/devicerepository.proto:
```diff
@@ -476,6 +484,24 @@ message MessagePayloadEncoder {
+  // Region.
+  string region = 5;
+}
+
```
This message is not required as an API since we don't use it any RPC call. You've already defined a similar message in `pkg/devicerepository/store/remote/schema.go` which will be used to parse the YAML file.

vlasebian replied:
I removed the message. I also had to move the EndDeviceProfile struct in the store package because it's used by both the implementations of the store. There was already the old EndDeviceProfile struct, so I named the new one EndDeviceProfileV2.

I didn't modify the existing structure because none of the fields from the EndDeviceProfileV2 are common to the EndDeviceProfile. It didn't make sense to me to drag along all those other fields when none of them are needed for the provisioning info. I also renamed all the methods so it's more suggestive that they use the new structure. I'm not really sure if it's the right approach, let me know if there is a better way of doing this.

KrishnaIyer commented on pkg/devicerepository/store/remote/testdata/vendor/repo-example/index.yaml:
```diff
@@ -0,0 +1,17 @@
+# This example contains just one end device: windsensor. It is referenced here in the index.
+
+endDevices:
+  # Unique identifier of the end device (lowercase, alphanumeric with dashes, max 36 characters)
+  - windsensor # look in windsensor.yaml for the end device definition
+
+# The profileIDs is a distinct value for every unique profile listed in the vendor's folder.
+# This value can be freely issued by the vendor and is also used on the QR code for LoRaWAN devices, see
+# https://lora-alliance.org/wp-content/uploads/2020/11/TR005_LoRaWAN_Device_Identification_QR_Codes.pdf#page=8
+# It can either be a combo of device ID + hardware version + firmware version + region, or profile ID + codec ID
+# NOTE: The profileIDs is different from the vendorID.
+profileIDs:
+  '1':
+    endDeviceID: 'windsensor'
+    firmwareVersion: '1.0'
+    hardwareVersion: '1.0'
+    region: 'EU863-870'
```
Why are these files here? How are other tests done without these files?

vlasebian replied:
I added a new vendor to have an example with end device profiles in the vendor index. The other examples have end device profiles embedded in the end device model only. Along with this file, the device and profile files are needed too as the code checks for their existence.

I removed the other unnecessary files.

KrishnaIyer reviewed (commented):
Only briefly looked at this. Look generally good.

johanstokking commented on api/ttn/lorawan/v3/devicerepository.proto:
```diff
@@ -386,6 +386,14 @@ message GetTemplateRequest {
 
+message GetEndDeviceProvisioningInfoRequest {
+  // VendorID managed by the LoRa Alliance, as defined in TR005.
+  uint32 vendor_id = 1 [(validate.rules).uint32 = {gte: 1}];
```
Reuse `LoRaAllianceProfileIdentifiers`

vlasebian replied:
Done.

johanstokking commented on api/ttn/lorawan/v3/devicerepository.proto:
```diff
@@ -476,6 +484,11 @@ message MessagePayloadEncoder {
   repeated Example examples = 4 [(validate.rules).repeated.max_items = 20];
 }
 
+message EndDeviceProvisioningInfo {
+  EndDeviceModel end_device_model = 1;
+  EndDeviceTemplate end_device_template = 2;
+}
```
1. I think this should be a response message; `GetEndDeviceProvisioningInfoResponse` unless it is an already-defined message. If you define a message just for a request or response, it should be a request or response message. This way we don't introduce messages with common names that are not commonly used except as request/response messages.
2. Isn't the available information more specific than the model? Don't we have `EndDeviceVersionIdentifiers` here?

vlasebian replied:
Item 1 is done.

vlasebian replied:
For item 2, I checked the network calls made by the console, EndDeviceVersionIdentifiers is the only thing that is needed for creating a device, as the end device ids are provided in the QR or have to be filled in:
```json
{
  "end_device": {
    "ids": {
      "join_eui": "0928745098237059",
      "dev_eui": "9283450982374509",
      "device_id": "dev2",
      "application_ids": {
        "application_id": "app1"
      }
    },
    "version_ids": {
      "brand_id": "the-things-industries",
      "model_id": "generic-node-sensor-edition",
      "hardware_version": "1.1",
      "firmware_version": "1.0",
      "band_id": "EU_863_870"
    },
    "network_server_address": "localhost",
    "application_server_address": "localhost",
    "join_server_address": "localhost"
  },
  "field_mask": {
    "paths": [
      "version_ids.brand_id",
      "version_ids.model_id",
      "version_ids.hardware_version",
      "version_ids.firmware_version",
      "version_ids.band_id",
      "network_server_address",
      "application_server_address",
      "join_server_address"
    ]
  }
}
```

However, if we want to reduce the number of calls done by the console, other details are needed about the device model, to show the model description, photo, product sheet url etc. that are needed in the device registration screen:

<img width="773" alt="Screenshot 2024-08-23 at 10 44 26" src="https://github.com/user-attachments/assets/aa4faed6-f80f-4eec-a995-30ea400a235e">

The full model seems relevant to have in the response. Let me know if I am missing something.

Also, regarding the template, does it make sense to have the full template in the response or is it okay to just have the end device? The message definition for the template looks like this:

```protobuf
message EndDeviceTemplate {
  EndDevice end_device = 1 [(validate.rules).message.required = true];
  google.protobuf.FieldMask field_mask = 2;
  string mapping_key = 3 [(validate.rules).string.max_len = 100];
}
```

So the EndDeviceTemplate is the EndDevice plus the field_mask and the mapping_key. Given that the GetEndDeviceProvisioningInfoResponse is contains two messages, the end_device_template and the end_device_model, a field_mask only for one of the messages seems unintuitive. What if we do:

```protobuf
message GetEndDeviceProvisioningInfoResponse {
  EndDeviceModel end_device_model = 1;
  EndDevice end_device = 2;
}
```

And if needed we can add a FieldMask for the full response, instead of having a field_mask for only one of the items? What do you think?

vlasebian replied:
@johanstokking what do you think about this?

johanstokking replied:
The API is not designed for the Console. If a presentation layer, like the Console, wants to present some additional data, it can request it using (existing) API.

In this case, my comment to return version identifiers is not only to return the most specific data possible for this endpoint, but also because the model does not contain the version identifiers. It contains the entire model, with potentially many firmware versions and regions. So clients cannot use it to create a new device. The goal is to know the version identifiers, so return that.

---

The way `EndDeviceTemplate` works is that it contains the `EndDevice` plus the field mask for the fields that are part of the template.

When setting a device to the NS, AS and JS registries, a field mask is required. So when creating a device, API clients (including the Console), need a field mask.

If we just return `EndDevice` with all kinds of fields set, the API client would have to figure out what fields are set to non-default values, and construct a field mask themselves. That is not API client friendly and may lead to all kinds of problems with default values etc.

So what API clients do, is they get the `EndDevice` plus the field mask of fields that are set. That is the template for creating an end device. Then, the version identifiers are set and added to the field mask. Most likely, the `EndDevice` is complete and the field mask is known, so it can be created.

vlasebian replied:
Okay, I understand now. I changed the response to use EndDeviceVersionIdentifiers instead of EndDeviceModel. I left the EndDeviceTemplate as is.

johanstokking reviewed (commented):
Only reviewed API

johanstokking commented on api/ttn/lorawan/v3/devicerepository.proto:
```diff
@@ -387,11 +387,7 @@ message GetTemplateRequest {
-
-  // ID of the LoRaWAN end device profile assigned by the vendor.
-  uint32 vendor_profile_id = 2;
+  LoRaAllianceProfileIdentifiers lora_alliance_profile_ids = 1;
```
Add validation that this is a required field

vlasebian replied:
Added!

johanstokking commented on pkg/devicerepository/grpc.go:
```diff
@@ -193,8 +193,8 @@ func (dr *DeviceRepository) GetEndDeviceProvisioningInfo(
 		store.GetProvisioningInfoRequest{
-			VendorID:        req.VendorId,
-			VendorProfileID: req.VendorProfileId,
+			VendorID:        req.LoraAllianceProfileIds.VendorId,
+			VendorProfileID: req.LoraAllianceProfileIds.VendorProfileId,
```
Otherwise this may panic

johanstokking commented on pkg/devicerepository/store/bleve/init.go:
```diff
@@ -71,6 +72,15 @@ type indexableProfile struct {
+	ProvisioningInfo *ttnpb.GetEndDeviceProvisioningInfoResponse
+
+	ProvisioningInfoJSON      string // *ttnpb.EndDeviceProvisioningInfo marshaled into string
+	VendorID, VendorProfileID string // stored separately to support queries
```
Not familiar with Bleve; should they be strings?

vlasebian replied:
They should be strings. I am also not that familiar with Bleve, but I checked the docs and Bleve doesn't seem to have any method of querying by an exact number (although it can query by a number range). Most of the query methods expect strings.

johanstokking commented on pkg/devicerepository/store/bleve/init.go:
```diff
@@ -203,11 +214,58 @@ func (c Config) Initialize(ctx context.Context, lorawanDevicesPath string, overw
 		}
 
+		// Add the end device provisioning info to the index.
+		endDeviceProfiles, err := s.GetEndDeviceProfilesV2(store.GetEndDeviceProfilesV2Request{BrandID: b.BrandID})
```
V2 is not idiomatic TTS. Use `ByXXX` to specialize an existing method name.

vlasebian replied:
After checking the comment below, I eventually renamed it to: `GetEndDeviceProfileIdentifiers`.

johanstokking commented on pkg/devicerepository/store/bleve/store.go:
```diff
@@ -243,6 +245,52 @@ func (s *bleveStore) GetTemplate(req *ttnpb.GetTemplateRequest, _ *store.EndDevi
+	if len(result.Hits) == 0 {
+		return nil, errProvisioningInfoNotFound.WithAttributes(
+			"vendor_id", req.VendorID,
+			"vendor_profile_id", req.VendorProfileID)
```
nit;

```suggestion
			"vendor_profile_id", req.VendorProfileID,
		)
```

vlasebian replied:
Fixed

johanstokking commented on pkg/devicerepository/store/profiles.go:
```diff
@@ -64,6 +64,20 @@ type EndDeviceProfile struct {
 
+// EndDeviceProfileV2 is the profile of a LoRaWAN end device that is mentioned in the vendor index file
+// instead of being embedded in the end device profile.
+type EndDeviceProfileV2 struct {
+	VendorProfileID string `yaml:"vendorProfileID"`
+	// First profile format.
+	EndDeviceID     string `yaml:"endDeviceID"`
+	FirmwareVersion string `yaml:"firmwareVersion"`
+	HardwareVersion string `yaml:"hardwareVersion"`
+	Region          string `yaml:"region"`
+	// Second profile format.
+	ID    string `yaml:"id"`
+	Codec string `yaml:"codec"`
+}
```
It's technically not a profile. These are identifiers to a profile.

The ID in the bottom is in fact a ProfileID.

vlasebian replied:
Renamed it.

johanstokking commented:
Looks great. I tested this locally. I had to do `tools/bin/mage dev:initDeviceRepo` to rebuild the index but otherwise it works.

Now, I see that the response is:

```json
{
    "end_device_version_ids": {
        "brand_id": "the-things-industries",
        "model_id": "generic-node-sensor-edition",
        "hardware_version": "1.1",
        "firmware_version": "1.0",
        "band_id": "EU_863_870"
    },
    "end_device_template": {
        "end_device": {
            "version_ids": {
                "brand_id": "the-things-industries",
                "model_id": "generic-node-sensor-edition",
                "hardware_version": "1.1",
                "firmware_version": "1.0",
                "band_id": "EU_863_870"
            },
            "lorawan_version": "MAC_V1_0_3",
            "lorawan_phy_version": "PHY_V1_0_3_REV_A",
            "supports_join": true,
            "mac_settings": {
                "supports_32_bit_f_cnt": true
            },
            "formatters": {
                "up_formatter": "FORMATTER_REPOSITORY",
                "down_formatter": "FORMATTER_REPOSITORY"
            }
        },
        "field_mask": {
            "paths": [
                "version_ids",
                "supports_join",
                "supports_class_b",
                "supports_class_c",
                "lorawan_version",
                "lorawan_phy_version",
                "formatters",
                "mac_settings.supports_32_bit_f_cnt"
            ]
        }
    }
}
```

So basically I wonder why we need to return the version identifiers if they're part of the template? I guess I didn't fully realize this before, but I think we can just return `EndDeviceTemplate` instead?

vlasebian commented:
I believe there already is an endpoint for that:
```
/dr/vendors/{end_device_profile_ids.vendor_id}/profiles/{end_device_profile_ids.vendor_profile_id}/template
```

But from what I understood at the beginning, the profiles can be specified in two places:
- Embedded in the device yaml
- Specified in the vendor index

This endpoint was for fetching the profiles specified in the vendor index.

johanstokking commented:
> But from what I understood at the beginning, the profiles can be specified in two places:
> 
>     * Embedded in the device yaml
> 
>     * Specified in the vendor index
> 
> 
> This endpoint was for fetching the profiles specified in the vendor index.

Ah I see now, thanks for clearing that up.

We changed the place where the vendor profile ID was defined. First, it was part of the profile itself. However, since the profile can be shared by multiple combinations of device IDs, firmware versions and hardware versions, we couldn't determine which device we're dealing with.

To overcome that limitation, we moved the vendor profile ID to a mapping at the vendor level. The other declaration no longer exists. To be backwards compatible, and to allow vendors to still use a vendor profile ID just for the profile and codec (that can be shared), we have this "secondary" use with references.

The only thing that is useful for us, however, are the "primary" identifiers as declared at the vendor-level mapping of profile ID to device ID, firmware version, hardware version and region.

If we can repurpose the existing API that you referenced to basically return the `EndDeviceTemplate` that we want, that would be most ideal.

vlasebian commented:
Sure, I can change the existing endpoint that fetches end device templates to fetch the vendor profile from the vendor index. Do you know if this endpoint is used somewhere else? I want to make sure it's a backward compatible change. Also, are we certain that profiles embedded in the end device model are not used anymore?

johanstokking commented:
> Sure, I can change the existing endpoint that fetches end device templates to fetch the vendor profile from the vendor index. Do you know if this endpoint is used somewhere else? I want to make sure it's a backward compatible change.

I'm not aware of use of it by us, maybe API clients, but as long as you keep the types the same I wouldn't expect any compatibility issues.

> Also, are we certain that profiles embedded in the end device model are not used anymore?

Yes we're sure because we removed them from the Device Repository. Adding them by mistake would cause validation to fail.

vlasebian commented:
Currently, these are the registered endpoints:
```protobuf
rpc GetTemplate(GetTemplateRequest) returns (EndDeviceTemplate) {
  option (google.api.http) = {
    get: "/dr/brands/{version_ids.brand_id}/models/{version_ids.model_id}/{version_ids.firmware_version}/{version_ids.band_id}/template"
    additional_bindings {get: "/dr/vendors/{end_device_profile_ids.vendor_id}/profiles/{end_device_profile_ids.vendor_profile_id}/template"}
    additional_bindings {get: "/dr/applications/{application_ids.application_id}/brands/{version_ids.brand_id}/models/{version_ids.model_id}/{version_ids.firmware_version}/{version_ids.band_id}/template"}
    additional_bindings {get: "/dr/applications/{application_ids.application_id}/vendors/{end_device_profile_ids.vendor_id}/profiles/{end_device_profile_ids.vendor_profile_id}/template"}
  };
}
```

And the GetTemplateRequest looks like this:
```protobuf
message GetTemplateRequest {
  // Identifiers to uniquely identify a LoRaWAN end device profile.
  message EndDeviceProfileIdentifiers {
    // VendorID managed by the LoRa Alliance, as defined in TR005.
    uint32 vendor_id = 1 [(validate.rules).uint32 = {gte: 1}];

    // ID of the LoRaWAN end device profile assigned by the vendor.
    uint32 vendor_profile_id = 2;
  }

  // Application identifiers.
  ApplicationIdentifiers application_ids = 1 [deprecated = true];

  // End device version information.
  // Use either EndDeviceVersionIdentifiers or EndDeviceProfileIdentifiers.
  EndDeviceVersionIdentifiers version_ids = 2;

  // End device profile identifiers.
  EndDeviceProfileIdentifiers end_device_profile_ids = 3;
}
```

Should the bindings that use something other than vendor_id and vendor_profile_id still be available? E.g. 
```
"/dr/brands/{version_ids.brand_id}/models/{version_ids.model_id}/{version_ids.firmware_version}/{version_ids.band_id}/template"
"/dr/applications/{application_ids.application_id}/brands/{version_ids.brand_id}/models/{version_ids.model_id}/{version_ids.firmware_version}/{version_ids.band_id}/template"
```

Or these should be deprecated along with the version_ids field in the GetTemplateRequest?

johanstokking commented:
Yes, keep them available, as we provide API compatibility until we release version 4.

Commit:
dr: Add new vendor in testdata

Commit:
dr: Update tests to include new test vendor

Commit:
dr: Update get template to fetch profiles from vendor index

Commit:
dr: Drop parameter in GetTemplates

Commit:
dr: Update api definition and translations

vlasebian commented:
I dropped all the changes related to the GetProvisioningInfo endpoint and started fresh. These are the changes that I did:
- Dropped the profiles index from the bleve store (it was used only to fetch templates)
- Added a new index for templates in the bleve store, so templates are returned directly instead of fetching a profile and then building the template
- Updated the reading of profiles from the device repository to account for the profile identifiers in the vendor index too. This is how reading end device profiles works now:
  - All the profiles are read as previously (all the device models are checked for profile identifiers, then the profiles are read from their own yaml files)
  - The vendor index is checked for any profile identifiers. For each profile identifier, the profile corresponding to that identifier is enriched with the vendor profile id. This is needed for the bleve index to be able to search using vendor id + vendor profile id.
- Dropped the profile parameter from the GetTemplate method in the store interface as it's not needed anymore
- Updated tests, fixed linter warnings

Let me know if this is okay now.

johanstokking reviewed (approved):
Great, LGTM

Please update the changelog

KrishnaIyer commented on pkg/devicerepository/store/bleve/bleve.go:
```diff
@@ -29,12 +31,12 @@ import (
-	// defaultTimeout is the timeout when trying to open the index. This is to avoid
-	// blocking on the index open call, which hangs indefinitely if the index is
-	// already in use by a different process.
+	// Timeout for opening the index. This is to avoid blocking on the index
```
Const, variable and function comments always start with the name first; https://go.dev/doc/comment
```
// defaultTimeout is the timeout for ...
```

vlasebian replied:
I got an error from the linter, it had the name initially.

<img width="597" alt="Screenshot 2024-09-10 at 17 52 08" src="https://github.com/user-attachments/assets/0b2c4ef5-fb1b-4765-965c-677078398a12">

I'll check if there's a way to disable that check for this kind of comments.

johanstokking replied:
Replace the first letter with a capital, no other change needed.

It seems like the rule for starting comments with the symbol name only applies to exposed stuff. Makes sense.

vlasebian replied:
done

KrishnaIyer commented on pkg/devicerepository/store/bleve/bleve.go:
```diff
@@ -29,12 +31,12 @@ import (
 	defaultTimeout = 5 * time.Second
 
-	// cacheSize is the size of the cache for brands and models.
+	// Size of the cache for brands and models.
```
Idem

vlasebian replied:
done

KrishnaIyer commented on pkg/devicerepository/store/bleve/store.go:
```diff
@@ -176,71 +188,82 @@ func (s *bleveStore) GetModels(req store.GetModelsRequest) (*store.GetModelsResp
+	)
+	errMissingProfileIdentifier = errors.DefineInvalidArgument(
+		"missing_profile_identifier",
+		"both vendor ID and vendor profile ID must be provided",
```
`errMissingProfileIdentifiers`

vlasebian replied:
done

KrishnaIyer commented on pkg/devicerepository/store/bleve/store.go:
```diff
@@ -176,71 +188,82 @@ func (s *bleveStore) GetModels(req store.GetModelsRequest) (*store.GetModelsResp
+
+// GetTemplate retrieves an end device template for an end device definition.
+func (s *bleveStore) GetTemplate(req *ttnpb.GetTemplateRequest) (*ttnpb.EndDeviceTemplate, error) {
+	profileIDs, versionIDs := req.GetEndDeviceProfileIds(), req.GetVersionIds()
```
These `versionIDs` don't seem to be used anywhere.

vlasebian replied:
The bleve index for templates uses only the profileIDs. In case no profileIDs are present, the index can't be used and the template fetch falls back on the remote implementation of the store where the versionIDs are used. I could check only for profileIDs here, in the bleve implementation of the GetTemplates and for versionIDs in the fallback method, in the remote implementation of the method.

vlasebian replied:
done

KrishnaIyer commented on pkg/devicerepository/store/remote/remote.go:
```diff
@@ -315,7 +410,58 @@ func (s *remoteStore) GetTemplate(req *ttnpb.GetTemplateRequest, profile *store.
+	ids *ttnpb.GetTemplateRequest_EndDeviceProfileIdentifiers,
+) (*ttnpb.EndDeviceVersionIdentifiers, error) {
+	if ids.VendorProfileId == 0 || ids.VendorId == 0 {
+		return nil, errors.New("vendor_profile_id and vendor_id must be set")
```
Please define a named error here.

vlasebian replied:
done

KrishnaIyer commented on pkg/devicerepository/store/remote/remote.go:
```diff
@@ -315,7 +410,58 @@ func (s *remoteStore) GetTemplate(req *ttnpb.GetTemplateRequest, profile *store.
+		return nil, errBrandWithVendorIDNotFound.WithAttributes("vendor_id", ids.VendorId)
+	}
+	endDeviceProfilesIdentifiers, err := s.GetEndDeviceProfileIdentifiers(
+		store.GetEndDeviceProfileIdentifiersRequest{BrandID: brandID})
```
nit: Add a new line before `)`.

vlasebian replied:
done

KrishnaIyer reviewed (commented):
Generally looks pretty good. A few comments.
